accepted/0000-caasp-nodes.md

```diff
@@ -0,0 +1,480 @@
+
+The patch can be upstreamed to Salt.
+
+The workflow for the user is then unchanged from the "usual" operation (before this RFC): issue any action (that is not forbidden) on the CaaS Platform node. The Salt formula to disable the blackout can only be used for troubleshooting.
+
+If a user schedules a forbidden action (not allowed on the `minion_blackout_whitelist`) the error message will be the same as the one described in the previous section.
+
+## Future work
+
+### Reboot action
+
+When scheduling a system reboot action from Uyuni/SUSE Manager, if the targeted minion is a CaaS Platform node, the action should be changed.
+We can make use of `kured` by simply changing the action to just `touch /var/run/reboot-required`.
+
+The related action are in `ActionChainManager#scheduleRebootAction`, `ActionChainManager#scheduleRebootActions`, and the XMLRPC action `SystemHandler#scheduleReboot`.
+
+### Package locking at `zypper` level
+
+Package locking of the forbidden packages is not an option: in that case, `skuba-update` is not able to patch the forbidden packages.
+
+While still maintaining minions in the blackout, the idea is to coordinate with the CaaS Platform team and:
+
+1. Package lock at the `zypper` level all the forbidden packages upon registration of a CaaS Platform node to Uyuni/SUSE Manager
+2. Modify `skuba-update` to release the package locks
+3. Unchanged `skuba-update` business
+4. `skuba-update` restores the package locks that were removed
+
+In this way, all forbidden packages operations can be allowed at the Salt level _and_ the Uyuni/SUSE Manager level: every operation that deals with the forbidden package will fail.
+
+There has been a discussion with the CaaS Platform team and the package locks were ultimately rejected during CaaS Platform design because package locks should be owned by the user, not by the system.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+<!-- Why should we **not** do this?
+
+  * obscure corner cases
+  * will it impact performance?
+  * what other parts of the product will be affected?
+  * will the solution be hard to maintain in the future? -->
+
+The user can still break its cluster in any way - but with this RFC we are at least offering a way to protect the user and the cluster.
+No messages will be displayed in the UI of Uyuni/SUSE Manager until the final failure of the forbidden action scheduled.
+
+# Alternatives
+[alternatives]: #alternatives
+
+<!-- - What other designs/options have been considered?
+- What is the impact of not doing this? -->
+
+### Disallow forbidden actions at both levels
+
+As an alternative to "Package locking at `zypper` level", we can consider adding another level of granularity: filter operations depending on the packages involved (e.g. forbid install of a package when the package is `kubernetes-kubeadm`).
+The idea is that:
+
+* At the Salt level: when the user tries to issue a command (e.g. `pkg.upgrade`) involving a forbidden package, the command will fail.
```

Comment:
I think this would require some more changes to Salt. We could avoid this by using package locking on the forbidden packages + minion `blackout` with `blackout_whitelist`.

Reply:
Isn't that covered in https://github.com/uyuni-project/uyuni-rfc/pull/26/files#diff-014d415bc2076bf85e73014398351426R289 ?

Reply:
Yes. I'm just suggesting that disallowing actions might not be robust enough and it might be turn out to be too complex, as you said.
The two options could be combined:
1. At the SUMA level prevent actions on forbidden packages
2. Forbid some Salt states and modules using the `blackout` mechanism (e.g. reboot) but allow states that upgrade/install packages. No need for more fine grained filtering at this level (e.g. no need to forbid certain packages)
3. Lock packages to prevent accidental changes from the Salt CLI or from the system CLI.